src/datachain/dataset.py

```diff
@@ -43,25 +43,58 @@
 StorageURI = NewType("StorageURI", str)
 
 
-def parse_dataset_uri(uri: str) -> tuple[str, Optional[str]]:
+def parse_dataset_uri(uri: str) -> tuple[str, str, str, str | None]:
     """
-    Parse dataser uri to extract name and version out of it (if version is defined)
-    Example:
-        Input: ds://zalando@v3.0.1
-        Output: (zalando, 3.0.1)
+    Parse a dataset URI of the form:
+
+        ds://<namespace>.<project>.<name>[@v<semver>]
+
+    Components:
+    - `ds://`        : required prefix identifying dataset URIs.
+    - `namespace`    : required namespace, may start with '@' (e.g., "@user").
+    - `project`      : required project name inside the namespace.
+    - `name`         : required dataset name.
+    - `@v<semver>`   : optional version suffix. Must start with '@v' and
+                       be a semantic version string MAJOR.MINOR.PATCH
+                       (e.g., "1.0.4").
+
+    Returns:
+        tuple[str, str, str, str | None]:
+            (namespace, project, name, version) where version is None
+            if not provided.
+
+    Raises:
+        ValueError: if the URI does not start with 'ds://' or does not
+                    match the expected format.
     """
-    p = urlparse(uri)
-    if p.scheme != "ds":
-        raise Exception("Dataset uri should start with ds://")
-    s = p.netloc.split("@v")
-    name = s[0]
-    if len(s) == 1:
-        return name, None
-    if len(s) != 2:
-        raise Exception(
-            "Wrong dataset uri format, it should be: ds://<name>@v<version>"
-        )
-    return name, s[1]
+
+    if not uri.startswith("ds://"):
+        raise ValueError(f"Invalid dataset URI: {uri}")
+
+    body = uri[len("ds://") :]
+
+    pattern = re.compile(
+        r"""
+        ^(?P<namespace>@?\w+)      # namespace, may start with '@'
+        \. (?P<project>\w+)        # project
```

Comment:
I bet we do allow `@` in project name too. Let's check? 🙏

Reply:
Yes, we do: https://github.com/iterative/datachain/blob/main/src/datachain/project.py#L10